File: torchard/tui/views/adopt_session.py
"""Adopt an unmanaged tmux session — associate it with a repo and base branch."""

from __future__ import annotations

from pathlib import Path

from textual.app import ComposeResult
from textual.binding import Binding
from textual.containers import Vertical
from textual.screen import Screen
from textual.widgets import Footer, Input, Label, ListItem, ListView, Static

from torchard.core.db import get_repos
from torchard.core.git import GitError, list_branches
from torchard.core.manager import Manager
from torchard.core.models import Repo
from torchard.tui.utils import safe_id


class AdoptSessionScreen(Screen):
    """Two-step wizard: pick a repo, pick a base branch, then adopt the session."""

    BINDINGS = [
        Binding("escape", "go_back", "Back", show=True),
    ]

    DEFAULT_CSS = """
    AdoptSessionScreen {
        align: center middle;
    }
    #adopt-container {
        width: 90;
        height: auto;
        max-height: 50;
        border: solid $accent;
        padding: 1 2;
    }
    #adopt-title {
        text-align: center;
        color: $accent;
        margin-bottom: 1;
        text-style: bold;
    }
    #adopt-hint {
        color: $text-muted;
        margin-bottom: 1;
        text-align: center;
    }
    #adopt-filter {
        margin-bottom: 1;
    }
    #adopt-list {
        height: 18;
        border: solid $panel;
    }
    #adopt-error {
        color: $error;
        margin-top: 1;
        text-align: center;
    }
    """

    def __init__(self, manager: Manager, session_name: str) -> None:
        super().__init__()
        self._manager = manager
        self._session_name = session_name
        self._step = 1  # 1 = repo, 2 = branch
        self._selected_repo: Repo | None = None
        self._repos: list[Repo] = []
        self._branches: list[str] = []
        self._awaiting_repo_path = False
        self._render_seq = 0  # uniquify widget IDs across re-renders
        self._id_to_branch: dict[str, str] = {}  # widget id -> actual branch name

    def compose(self) -> ComposeResult:
        with Vertical(id="adopt-container"):
            yield Static("", id="adopt-title")
            yield Static("", id="adopt-hint")
            yield Input(placeholder="Filter…", id="adopt-filter")
            yield ListView(id="adopt-list")
            yield Static("", id="adopt-error")
        yield Footer()

    def on_mount(self) -> None:
        self._render_step()

    def _set_title(self, text: str) -> None:
        self.query_one("#adopt-title", Static).update(text)

    def _set_hint(self, text: str) -> None:
        self.query_one("#adopt-hint", Static).update(text)

    def _set_error(self, text: str) -> None:
        self.query_one("#adopt-error", Static).update(text)

    def _render_step(self) -> None:
        self._awaiting_repo_path = False
        self._set_error("")
        if self._step == 1:
            self._render_repo_step()
        elif self._step == 2:
            self._render_branch_step()

    def _render_repo_step(self) -> None:
        self._set_title(f"Adopt '{self._session_name}' — Select Repository")
        self._set_hint("Type to filter. Enter to select.")
        fi = self.query_one("#adopt-filter", Input)
        fi.placeholder = "Filter repos…"
        fi.value = ""
        self._repos = get_repos(self._manager._conn)
        self._populate_repo_list(self._repos)
        fi.focus()

    def _render_branch_step(self) -> None:
        assert self._selected_repo is not None
        self._set_title(f"Adopt '{self._session_name}' — Select Branch  [dim]({self._selected_repo.name})[/dim]")
        self._set_hint("Pick the branch new worktrees will branch from.")
        fi = self.query_one("#adopt-filter", Input)
        fi.placeholder = "Filter or type a new branch…"
        fi.value = ""
        try:
            self._branches = list_branches(self._selected_repo.path)
        except GitError as exc:
            self._branches = []
            self._set_error(str(exc))
        self._populate_branch_list(self._branches, "")
        fi.focus()

    def _populate_repo_list(self, repos: list[Repo]) -> None:
        self._render_seq += 1
        seq = self._render_seq
        lv = self.query_one("#adopt-list", ListView)
        lv.clear()
        for repo in repos:
            lv.append(ListItem(Label(f"[bold]{repo.name}[/bold]  [dim]{repo.path}[/dim]"), id=f"repo-{repo.id}-{seq}"))
        lv.append(ListItem(Label("[green]+ Add new repo path…[/green]"), id=f"add-repo-{seq}"))

    def _populate_branch_list(self, branches: list[str], query: str) -> None:
        self._render_seq += 1
        seq = self._render_seq
        self._id_to_branch.clear()
        lv = self.query_one("#adopt-list", ListView)
        lv.clear()
        for branch in branches:
            widget_id = f"branch-{safe_id(branch)}-{seq}"
            self._id_to_branch[widget_id] = branch
            lv.append(ListItem(Label(branch), id=widget_id))
        if query and query not in branches:
            lv.append(ListItem(Label(f"[green]+ New branch: [bold]{query}[/bold][/green]"), id=f"new-branch-{seq}"))

    def on_input_changed(self, event: Input.Changed) -> None:
        if event.input.id != "adopt-filter" or self._awaiting_repo_path:
            return
        query = event.value.lower()
        if self._step == 1:
            filtered = [r for r in self._repos if query in r.name.lower() or query in r.path.lower()]
            self._populate_repo_list(filtered)
        elif self._step == 2:
            filtered = [b for b in self._branches if query in b.lower()]
            self._populate_branch_list(filtered, event.value)

    def on_input_submitted(self, event: Input.Submitted) -> None:
        if event.input.id != "adopt-filter":
            return
        if self._awaiting_repo_path:
            self._finish_add_repo(event.value)
        else:
            self._confirm_list_selection(typed_value=event.value)

    def on_list_view_selected(self, event: ListView.Selected) -> None:
        item_id = event.item.id
        if self._step == 1:
            self._select_repo_item(item_id)
        elif self._step == 2:
            self._select_branch_item(item_id)

    def _confirm_list_selection(self, typed_value: str) -> None:
        lv = self.query_one("#adopt-list", ListView)
        highlighted = lv.highlighted_child
        if highlighted is None and len(lv.children) > 0:
            highlighted = lv.children[0]
        if highlighted is not None:
            item_id = highlighted.id
            if self._step == 1:
                self._select_repo_item(item_id)
            elif self._step == 2:
                self._select_branch_item(item_id)
        elif self._step == 2 and typed_value:
            self._adopt(typed_value)

    def _select_repo_item(self, item_id: str | None) -> None:
        if item_id and item_id.startswith("add-repo"):
            self._enter_repo_path_mode()
            return
        if item_id and item_id.startswith("repo-"):
            # id format: repo-{id}-{seq}
            parts = item_id.split("-")
            repo_id = int(parts[1])
            repo = next((r for r in self._repos if r.id == repo_id), None)
            if repo is not None:
                self._selected_repo = repo
                self._step = 2
                self._render_step()

    def _select_branch_item(self, item_id: str | None) -> None:
        if item_id and item_id.startswith("new-branch"):
            typed = self.query_one("#adopt-filter", Input).value
            if typed:
                self._adopt(typed)
            return
        if item_id and item_id in self._id_to_branch:
            self._adopt(self._id_to_branch[item_id])

    def _adopt(self, base_branch: str) -> None:
        assert self._selected_repo is not None
        try:
            self._manager.adopt_session(
                session_name=self._session_name,
                repo_path=self._selected_repo.path,
                base_branch=base_branch,
            )
        except Exception as exc:
            self._set_error(f"Error: {exc}")
            return
        self.app.pop_screen()

    def _enter_repo_path_mode(self) -> None:
        self._awaiting_repo_path = True
        self._set_hint("Enter the full path to a git repo, then press Enter.")
        fi = self.query_one("#adopt-filter", Input)
        fi.placeholder = "e.g. /home/you/dev/myproject…"
        fi.value = ""
        fi.focus()

    def _finish_add_repo(self, path_str: str) -> None:
        self._awaiting_repo_path = False
        path = Path(path_str.strip()).expanduser().resolve()
        if not path.is_dir() or not (path / ".git").exists():
            self._set_error(f"'{path}' is not a git repository.")
            fi = self.query_one("#adopt-filter", Input)
            fi.placeholder = "Filter repos…"
            fi.value = ""
            self._set_hint("Type to filter. Enter to select.")
            return

        class _AdhocRepo:
            def __init__(self, p: Path) -> None:
                self.path = str(p)
                self.name = p.name
                self.id = None

        self._selected_repo = _AdhocRepo(path)  # type: ignore[assignment]
        self._step = 2
        self._render_step()

    def action_go_back(self) -> None:
        if self._awaiting_repo_path:
            self._awaiting_repo_path = False
            fi = self.query_one("#adopt-filter", Input)
            fi.placeholder = "Filter repos…"
            fi.value = ""
            self._populate_repo_list(self._repos)
            self._set_hint("Type to filter. Enter to select.")
            self._set_error("")
            return
        if self._step == 1:
            self.app.pop_screen()
        else:
            self._step = 1
            self._selected_repo = None
            self._render_step()

File: src/tui/adopt_session.rs
// torchard-rs/src/tui/adopt_session.rs

use crossterm::event::{Event, KeyCode, KeyEvent, KeyEventKind};
use ratatui::prelude::*;
use ratatui::widgets::{Block, Borders, List, ListItem, ListState, Paragraph};

use crate::git;
use crate::manager::Manager;
use crate::models::Repo;
use super::rename::{input_handle_key, render_input, render_modal_box};
use super::{ScreenAction, ScreenBehavior};
use super::theme;

// ---------------------------------------------------------------------------
// Wizard step
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Copy, PartialEq)]
enum Step {
    PickRepo,
    PickBranch,
}

// ---------------------------------------------------------------------------
// A repo selection (either from DB or ad-hoc path entry)
// ---------------------------------------------------------------------------

#[derive(Clone)]
struct SelectedRepo {
    name: String,
    path: String,
}

// ---------------------------------------------------------------------------
// AdoptSessionScreen
// ---------------------------------------------------------------------------

pub struct AdoptSessionScreen {
    session_name: String,
    step: Step,

    // Repo step
    repos: Vec<Repo>,
    filtered_repos: Vec<usize>, // indices into repos
    repo_list_state: ListState,
    awaiting_repo_path: bool,

    // Branch step
    branches: Vec<String>,
    filtered_branches: Vec<String>,
    branch_list_state: ListState,

    // Selection
    selected_repo: Option<SelectedRepo>,

    // Shared filter input
    filter_input: String,
    filter_cursor: usize,

    // Error display
    error: String,
}

impl AdoptSessionScreen {
    pub fn new(manager: &Manager, session_name: String) -> Self {
        let repos = manager.get_repos();
        let filtered_repos: Vec<usize> = (0..repos.len()).collect();
        let mut repo_list_state = ListState::default();
        if !filtered_repos.is_empty() {
            repo_list_state.select(Some(0));
        }
        Self {
            session_name,
            step: Step::PickRepo,
            repos,
            filtered_repos,
            repo_list_state,
            awaiting_repo_path: false,
            branches: Vec::new(),
            filtered_branches: Vec::new(),
            branch_list_state: ListState::default(),
            selected_repo: None,
            filter_input: String::new(),
            filter_cursor: 0,
            error: String::new(),
        }
    }

    // ------------------------------------------------------------------
    // Step transitions
    // ------------------------------------------------------------------

    fn go_to_repo_step(&mut self) {
        self.step = Step::PickRepo;
        self.selected_repo = None;
        self.filter_input.clear();
        self.filter_cursor = 0;
        self.error.clear();
        self.awaiting_repo_path = false;
        self.filtered_repos = (0..self.repos.len()).collect();
        self.repo_list_state = ListState::default();
        if !self.filtered_repos.is_empty() {
            self.repo_list_state.select(Some(0));
        }
    }

    fn go_to_branch_step(&mut self) {
        self.step = Step::PickBranch;
        self.filter_input.clear();
        self.filter_cursor = 0;
        self.error.clear();

        let repo = self.selected_repo.as_ref().unwrap();
        match git::list_branches(&repo.path) {
            Ok(branches) => {
                self.branches = branches;
            }
            Err(e) => {
                self.branches = Vec::new();
                self.error = e.to_string();
            }
        }
        self.filtered_branches = self.branches.clone();
        self.branch_list_state = ListState::default();
        if !self.filtered_branches.is_empty() {
            self.branch_list_state.select(Some(0));
        }
    }

    // ------------------------------------------------------------------
    // Path entry mode
    // ------------------------------------------------------------------

    fn enter_repo_path_mode(&mut self) {
        self.awaiting_repo_path = true;
        self.filter_input.clear();
        self.filter_cursor = 0;
        self.error.clear();
    }

    fn finish_add_repo(&mut self) -> ScreenAction {
        let path_str = self.filter_input.trim().to_string();
        let path = if path_str.starts_with('~') {
            let home = dirs::home_dir().unwrap_or_default();
            home.join(path_str.strip_prefix("~/").unwrap_or(&path_str[1..]))
        } else {
            std::path::PathBuf::from(&path_str)
        };
        let path = match path.canonicalize() {
            Ok(p) => p,
            Err(_) => path,
        };

        if !path.is_dir() || !path.join(".git").exists() {
            self.error = format!("'{}' is not a git repository.", path.display());
            self.awaiting_repo_path = false;
            self.filter_input.clear();
            self.filter_cursor = 0;
            return ScreenAction::None;
        }

        let name = path
            .file_name()
            .unwrap_or_default()
            .to_string_lossy()
            .to_string();
        let path_s = path.to_string_lossy().to_string();
        self.selected_repo = Some(SelectedRepo { name, path: path_s });
        self.awaiting_repo_path = false;
        self.go_to_branch_step();
        ScreenAction::None
    }

    // ------------------------------------------------------------------
    // Adopt
    // ------------------------------------------------------------------

    fn adopt(&mut self, base_branch: &str, manager: &mut Manager) -> ScreenAction {
        let repo = self.selected_repo.as_ref().unwrap();
        match std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
            manager.adopt_session(&self.session_name, &repo.path, base_branch)
        })) {
            Ok(_) => ScreenAction::Pop,
            Err(_) => {
                self.error = "Failed to adopt session.".to_string();
                ScreenAction::None
            }
        }
    }

    // ------------------------------------------------------------------
    // Filtering
    // ------------------------------------------------------------------

    fn apply_filter(&mut self) {
        let query = self.filter_input.to_lowercase();
        match self.step {
            Step::PickRepo => {
                self.filtered_repos = self
                    .repos
                    .iter()
                    .enumerate()
                    .filter(|(_, r)| {
                        r.name.to_lowercase().contains(&query)
                            || r.path.to_lowercase().contains(&query)
                    })
                    .map(|(i, _)| i)
                    .collect();
                self.repo_list_state = ListState::default();
                if !self.filtered_repos.is_empty() {
                    self.repo_list_state.select(Some(0));
                }
            }
            Step::PickBranch => {
                self.filtered_branches = self
                    .branches
                    .iter()
                    .filter(|b| b.to_lowercase().contains(&query))
                    .cloned()
                    .collect();
                self.branch_list_state = ListState::default();
                if !self.filtered_branches.is_empty() {
                    self.branch_list_state.select(Some(0));
                }
            }
        }
    }

    // ------------------------------------------------------------------
    // List navigation
    // ------------------------------------------------------------------

    fn list_len(&self) -> usize {
        match self.step {
            Step::PickRepo => self.filtered_repos.len() + 1, // +1 for "add repo"
            Step::PickBranch => {
                let extra = if !self.filter_input.is_empty()
                    && !self.filtered_branches.contains(&self.filter_input)
                {
                    1
                } else {
                    0
                };
                self.filtered_branches.len() + extra
            }
        }
    }

    fn selected_index(&self) -> Option<usize> {
        match self.step {
            Step::PickRepo => self.repo_list_state.selected(),
            Step::PickBranch => self.branch_list_state.selected(),
        }
    }

    fn select_index(&mut self, idx: usize) {
        match self.step {
            Step::PickRepo => self.repo_list_state.select(Some(idx)),
            Step::PickBranch => self.branch_list_state.select(Some(idx)),
        }
    }

    fn move_selection(&mut self, delta: i32) {
        let len = self.list_len();
        if len == 0 {
            return;
        }
        let current = self.selected_index().unwrap_or(0) as i32;
        let next = (current + delta).rem_euclid(len as i32) as usize;
        self.select_index(next);
    }

    // ------------------------------------------------------------------
    // Confirm selection (Enter)
    // ------------------------------------------------------------------

    fn confirm_selection(&mut self, manager: &mut Manager) -> ScreenAction {
        match self.step {
            Step::PickRepo => self.confirm_repo(manager),
            Step::PickBranch => self.confirm_branch(manager),
        }
    }

    fn confirm_repo(&mut self, _manager: &mut Manager) -> ScreenAction {
        let idx = match self.repo_list_state.selected() {
            Some(i) => i,
            None => {
                if self.filtered_repos.is_empty() {
                    self.enter_repo_path_mode();
                    return ScreenAction::None;
                }
                0
            }
        };
        if idx >= self.filtered_repos.len() {
            // "Add new repo path..." item
            self.enter_repo_path_mode();
            return ScreenAction::None;
        }
        let repo_idx = self.filtered_repos[idx];
        let repo = &self.repos[repo_idx];
        self.selected_repo = Some(SelectedRepo {
            name: repo.name.clone(),
            path: repo.path.clone(),
        });
        self.go_to_branch_step();
        ScreenAction::None
    }

    fn confirm_branch(&mut self, manager: &mut Manager) -> ScreenAction {
        let idx = self.branch_list_state.selected().unwrap_or(0);
        let branch = if idx < self.filtered_branches.len() {
            self.filtered_branches[idx].clone()
        } else {
            // "New branch" item — use the typed text
            if self.filter_input.is_empty() {
                return ScreenAction::None;
            }
            self.filter_input.clone()
        };
        self.adopt(&branch, manager)
    }

    // ------------------------------------------------------------------
    // Rendering helpers
    // ------------------------------------------------------------------

    fn title(&self) -> String {
        match self.step {
            Step::PickRepo => format!("Adopt '{}' — Select Repository", self.session_name),
            Step::PickBranch => {
                let repo_name = self.selected_repo.as_ref().map(|r| r.name.as_str()).unwrap_or("?");
                format!("Adopt '{}' — Select Branch ({})", self.session_name, repo_name)
            }
        }
    }

    fn hint(&self) -> &str {
        if self.awaiting_repo_path {
            return "Enter the full path to a git repo, then press Enter.";
        }
        match self.step {
            Step::PickRepo => "Type to filter. Enter to select. Escape to cancel.",
            Step::PickBranch => "Pick the branch new worktrees will branch from.",
        }
    }

    fn placeholder(&self) -> &str {
        if self.awaiting_repo_path {
            return "e.g. /home/you/dev/myproject";
        }
        match self.step {
            Step::PickRepo => "Filter repos...",
            Step::PickBranch => "Filter or type a new branch...",
        }
    }

    fn build_list_items(&self) -> Vec<ListItem<'_>> {
        match self.step {
            Step::PickRepo => {
                let mut items: Vec<ListItem> = self
                    .filtered_repos
                    .iter()
                    .map(|&i| {
                        let repo = &self.repos[i];
                        let line = Line::from(vec![
                            Span::styled(
                                repo.name.as_str(),
                                Style::default().fg(theme::TEXT).add_modifier(Modifier::BOLD),
                            ),
                            Span::styled("  ", Style::default()),
                            Span::styled(repo.path.as_str(), Style::default().fg(theme::TEXT_DIM)),
                        ]);
                        ListItem::new(line)
                    })
                    .collect();
                items.push(ListItem::new(Span::styled(
                    "+ Add new repo path...",
                    Style::default().fg(theme::GREEN),
                )));
                items
            }
            Step::PickBranch => {
                let mut items: Vec<ListItem> = self
                    .filtered_branches
                    .iter()
                    .map(|b| ListItem::new(Span::styled(b.as_str(), Style::default().fg(theme::TEXT))))
                    .collect();
                if !self.filter_input.is_empty()
                    && !self.filtered_branches.contains(&self.filter_input)
                {
                    let line = Line::from(vec![
                        Span::styled("+ New branch: ", Style::default().fg(theme::GREEN)),
                        Span::styled(
                            self.filter_input.as_str(),
                            Style::default().fg(theme::GREEN).add_modifier(Modifier::BOLD),
                        ),
                    ]);
                    items.push(ListItem::new(line));
                }
                items
            }
        }
    }
}

impl ScreenBehavior for AdoptSessionScreen {
    fn render(&self, f: &mut Frame, area: Rect, _manager: &Manager) {
        let title = self.title();
        let inner = render_modal_box(f, area, &title, 80, 28);

        let chunks = Layout::vertical([
            Constraint::Length(1), // hint
            Constraint::Length(1), // spacer
            Constraint::Length(1), // filter input
            Constraint::Length(1), // spacer
            Constraint::Min(4),    // list
            Constraint::Length(1), // error
        ])
        .split(inner);

        // Hint
        let hint = Paragraph::new(Span::styled(self.hint(), Style::default().fg(theme::TEXT_DIM)))
            .alignment(Alignment::Center);
        f.render_widget(hint, chunks[0]);

        // Filter input
        if self.filter_input.is_empty() && !self.awaiting_repo_path {
            let ph = Paragraph::new(Span::styled(
                self.placeholder(),
                Style::default().fg(Color::Rgb(0x66, 0x66, 0x66)),
            ));
            f.render_widget(ph, chunks[2]);
            let cursor_area = Rect { x: chunks[2].x, y: chunks[2].y, width: chunks[2].width, height: 1 };
            render_input(f, cursor_area, "", 0);
        } else {
            render_input(f, chunks[2], &self.filter_input, self.filter_cursor);
        }

        // List (only when not in path entry mode)
        if !self.awaiting_repo_path {
            let items = self.build_list_items();
            let list_state = match self.step {
                Step::PickRepo => &self.repo_list_state,
                Step::PickBranch => &self.branch_list_state,
            };

            let list = List::new(items)
                .block(
                    Block::default()
                        .borders(Borders::ALL)
                        .border_style(Style::default().fg(theme::ACCENT))
                        .style(Style::default().bg(theme::BG)),
                )
                .highlight_style(
                    Style::default()
                        .bg(theme::CURSOR_BG)
                        .add_modifier(Modifier::BOLD),
                )
                .highlight_symbol("> ");

            let mut state = list_state.clone();
            f.render_stateful_widget(list, chunks[4], &mut state);
        }

        // Error
        if !self.error.is_empty() {
            let err = Paragraph::new(Span::styled(&self.error[..], theme::style_error()))
                .alignment(Alignment::Center);
            f.render_widget(err, chunks[5]);
        }
    }

    fn handle_event(&mut self, event: &Event, manager: &mut Manager) -> ScreenAction {
        if let Event::Key(KeyEvent { code, kind: KeyEventKind::Press, modifiers, .. }) = event {
            match code {
                KeyCode::Esc => {
                    if self.awaiting_repo_path {
                        self.awaiting_repo_path = false;
                        self.filter_input.clear();
                        self.filter_cursor = 0;
                        self.error.clear();
                        self.filtered_repos = (0..self.repos.len()).collect();
                        self.repo_list_state = ListState::default();
                        if !self.filtered_repos.is_empty() {
                            self.repo_list_state.select(Some(0));
                        }
                        return ScreenAction::None;
                    }
                    match self.step {
                        Step::PickRepo => return ScreenAction::Pop,
                        Step::PickBranch => {
                            self.go_to_repo_step();
                            return ScreenAction::None;
                        }
                    }
                }
                KeyCode::Enter => {
                    if self.awaiting_repo_path {
                        return self.finish_add_repo();
                    }
                    return self.confirm_selection(manager);
                }
                KeyCode::Up => {
                    if !self.awaiting_repo_path {
                        self.move_selection(-1);
                    }
                    return ScreenAction::None;
                }
                KeyCode::Down => {
                    if !self.awaiting_repo_path {
                        self.move_selection(1);
                    }
                    return ScreenAction::None;
                }
                KeyCode::Tab => {
                    if !self.awaiting_repo_path {
                        self.move_selection(1);
                    }
                    return ScreenAction::None;
                }
                KeyCode::BackTab => {
                    if !self.awaiting_repo_path {
                        self.move_selection(-1);
                    }
                    return ScreenAction::None;
                }
                _ => {
                    input_handle_key(&mut self.filter_input, &mut self.filter_cursor, *code, *modifiers);
                    if !self.awaiting_repo_path {
                        self.apply_filter();
                    }
                    self.error.clear();
                }
            }
        }
        ScreenAction::None
    }

    fn is_modal(&self) -> bool {
        true
    }
}
